### PR TITLE
Handle V8 heap snapshots well + allow custom snapshot generation

### PR DESCRIPTION
**Pull request by stesie**
* this tries to detect whether V8 has been built with snapshot generation enabled
* ... and calls `v8::V8::InitializeExternalStartupData` as needed, if yes
* adds a new static method `V8Js::createSnapshot` which creates a new heap snapshot (blob returned to caller as a string)
* allows this snapshot blob to be provided to `V8Js::__construct` as new fifth argument

This refs issue #205

### CODE REVIEW

**Comment by virgofx** [comment]
@stesie I'm getting a `configure: error: Please provide V8 native blob as needed` during `./configure`.

**Comment by stesie** [comment]
Please provide the contents of one out file, e.g. `tests/array_access.out`

**Comment by virgofx** [comment]
Rebuilding again from fresh V8-5.0.104 with this PR. Error during configure:

```
...
checking for V8 version... 5.0.104
checking for libv8_libplatform.a... found in /usr
checking whether V8 requires startup data... yes
checking for natives_blob.bin... not found
configure: error: Please provide V8 native blob as needed
```

**Comment by virgofx** [comment]
Rebuilt against the PR and was getting this error.

```
checking for V8 version... 5.0.104
checking for libv8_libplatform.a... found in /usr
checking whether V8 requires startup data... yes
checking for natives_blob.bin... not found
configure: error: Please provide V8 native blob as needed
```

Where can I provide the path to the native blob (Currently located in `/usr/share/v8/out/native/natives_blob.bin`) ?

**Comment by virgofx** [comment]
@stesie I'm still getting this `Invariant Violation: Trying to release an instance into a pool of a different type`

**Comment by pinepain** [comment]
@virgofx As to `natives_blob.bin` problem, make sure you you build with snapshot enabled but without external startup data:

from my PPA for libv8 (https://launchpad.net/~pinepain/+archive/ubuntu/libv8-4.10):

```sh
GYPFLAGS += -Dv8_use_snapshot=true -Dv8_use_external_startup_data=0

DEB_MAKE_EXTRA_ARGS += library=shared snapshot=on i18nsupport=on
```

same for homebrew recipe, if you are on OS X and use homebrew - ping me and I'll share recipe.

If you are on ubuntu feels free to use my PPA, it but pay attention while I update it time to time and due to v8 API BC incompatible changes some update may be incompatible with v8js. It is less likely for 4.10 branch while I have in plan to start building 5.x soon.

@stesie ping me so maybe we'll try to figure out about providing libv8 on ubuntu that will fits v8js versions, currently I build it for my own php-v8 extension.

**Comment by stesie** [comment]
@virgofx yes, `configure` currently checks for the external snapshots in `/usr/lib` (or whereever else you put the `libv8.so` file). However 

- [ ] `/usr/share/v8/...` would be a better fit, check that as well
- [ ] update `README.Linux.md` file to reflect the option

further TODOs left

- [ ] add phpt test case for `createSnapshot` functionality
- [ ] make sure `createSnapshot` works with snapshots enabled, yet `v8_use_external_startup_data=0`

The *Invariant Violation: Trying to release an instance into a pool of a different type* exception is thrown by React. So actually you already have it kindof working. I haven't used React before so I'm unsure what really causes it. To get it working I initially just commented it out, then noticed further "real" problems and once those were sorted out I could savely comment it back in and thigs seem to work fine.

Actually I had to fix two problems: I had `print` statements lingering around and `Math.random` caused me some trouble.

File: v8js_class.cc [hunk 1]
```diff
@@ -1063,6 +1081,40 @@ static PHP_METHOD(V8Js, getExtensions)
...
+
+	RETVAL_STRINGL(snapshot_blob.data, snapshot_blob.raw_size, 1);
+	delete[] snapshot_blob.data;
+}
+/* }}} */
+#endif  /* PHP_V8_USE_EXTERNAL_STARTUP_DATA */
+
+
 /* {{{ arginfo */
 ZEND_BEGIN_ARG_INFO_EX(arginfo_v8js_construct, 0, 0, 0)
```
Comment (virgofx) [inline v8js_class.cc]: I notice there is no argument defined for the new snapshot blob which is defined as a parameter above?

**Comment by virgofx** [comment]
Thanks for the help guys. I have managed to get things worked externally but not able to use the `createSnapshot` functionality.  Let me know if you have any thoughts/comment on my process below.

#### Environment
1. Compile V8   ✔
    ```
cd /usr/share
rm -rf v8
rm -rf .gclient*
fetch v8
cd v8
git checkout 5.0.105
gclient sync
make native library=shared snapshot=on disassembler=off -j2
mkdir -p /usr/lib /usr/include
cp out/native/lib.target/lib*.so /usr/lib/
cp -R include/* /usr/include
echo -e "create /usr/lib/libv8_libplatform.a\naddlib out/native/obj.target/tools/gyp/libv8_libplatform.a\nsave\nend" | sudo ar -M
```

    > Have not tested with `v8_use_external_startup_data=0` as @pinepain recommends.
1. Copy `natives_blob.bin` and `snapshot_blob.bin` into `/usr/lib` ✔

1. Clone V8JS - custom snapshots ✔

1. Make ✔
    ```
phpize
./configure
make
make test
make install
service php5-fpm restart
```

**Extensions loaded, PHP working** ✔

====

#### Custom Snapshot Data

##### Manually
Create snapshot via `mksnapshot` and then overwrite the custom_snapshot.bin which gets loaded on startup.
```
sudo /usr/share/v8/out/native/mksnapshot --startup_blob=custom_snapshot.bin /tmp/doublify.js
sudo cp /usr/share/v8/out/native/custom_snapshot.bin /usr/lib/snapshot_blob.bin
```

##### Via PHP

Can't get anything to work with the API?

1. Create Snapshot ✔

    ```
$v8js = new V8Js();
$snapshotBlob = $v8js->createSnapshot(file_get_contents('/tmp/doublify.js'));
```

1. Trying to load via constructor doesn't seem to work.

**Comment by virgofx** [comment]
@stesie Manually compiling, I can at least get back to the runtime Invariant issue associated with React + snapshots.  What's weird is that it works using the same concatenated data via execute string; however, not as a snapshot.

If you do have an exact binary of the included script you used for working React server side, let me know. I'm currently using:

```
var global = global || this, self = self || this, window = window || this;
// [+] react.js  (0.14.7 from bower)
// [+] react-dom-server.js (0.14.7 from bower)
```

**Comment by virgofx** [comment]
@stesie Removing the Math.random()'s did fix the invariant problem and now I have results. Will post on main issue. I confirm similar results to you!

**Comment by virgofx** [comment]
@stesie Do you know if the Math.random() issue is V8 related or V8JS related? I originally thought it was React related; however, I can reproduce the problem in V8JS by attempting to use Math.random() outside the global scope. 

e.g.
```js

random1 = Math.random();

function getRandom() {
    return Math.random();
}
```

```js
print(random1); // Works
print(getRandom()); // Bugs out
```

**Comment by stesie** [comment]
@virgofx very cool that it finally works for you too! :-)

Regarding the arginfo, that one is "just" for PHP's reflection stuff, hence yes, 

- [ ] fix arginfo of `V8Js::__construct`

... but it is *not* relevant to execution.  Have you re-tried snapshot passing via `__construct`?

BTW the `createSnapshot` is a static method, so you don't have to (and actually shouldn't) invoke it on an object instance but rather on the class itself.

```console
stesie@hahnschaaf:~/Projekte/v8js$ php5 -n -d extension_dir=./modules -d extension=v8js.so -d extension=readline.so -a
Interactive mode enabled

php > $snapshotBlob = V8Js::createSnapshot(file_get_contents('doublify.js'));
php > $v8 = new V8Js('PHP', [], [], true, $snapshotBlob);
php > $v8->executeString('print(doublify(23));');
46
```

Regarding `Math.random` I suppose that is V8's fault - I'm uncertain if it is expected behaviour or buggy so. Adding to your observations it is even more strange that your `getRandom()` function works as expected, if used within embedded script code, ... it just starts failing if used from outside of the snapshot.

... and to show it's not related to V8Js, you can simply reproduce it with V8's own shell (and no V8Js code at all):

Create a file test.js:

```js
function getRandom() {
  return Math.random();
}

var randomValue = Math.random();
var randomValueFromFunction = getRandom();
```

create a snapshot

```console
stesie@hahnschaaf:~/Projekte/v8/out/native$ ./mksnapshot --startup_blob=snapshot_blob.bin test.js                                                      
Embedding extra script: test.js
```

start a shell that uses the startup_blob

```console
stesie@hahnschaaf:~/Projekte/v8/out/native$ ./shell 
V8 version 5.1.53 [sample shell]
> print(randomValue);
0.47515791309267663
> print(randomValueFromFunction);
0.05808531226718627
> print(getRandom());
<embedded script>:6: TypeError: Cannot read property '4' of undefined
  return Math.random();
              ^
TypeError: Cannot read property '4' of undefined
    at Object.random (native)
    at getRandom (<embedded script>:6:15)
    at (shell):1:7
```

**Comment by virgofx** [comment]
@stesie Just some more notes for Math.random() - It appears to error consistently when the Math.random() needs to generate a value at runtime instead of compile-time.

**Comment by virgofx** [comment]
Submitted an issue over at [V8 issue tracker.](https://bugs.chromium.org/p/v8/issues/detail?id=4810) 

For now anyone who is following this can semi-polyfill by creating snapshot with a function to handle runtime generation by simply overwriting the `Math.random` definition.

```js
Math.random = function () { 
    // Quick, simple version.
    return Date.now() / 3000000000000; 
}
```

**Comment by stesie** [comment]
another work-around: overwrite `Math.random` with a call to PHP and get random numbers from there:

```php
    $this->v8 = new V8Js('PHP', [], [], true, $blob);
    $this->v8->rand = function() { return mt_rand() / mt_getrandmax(); };
    $this->v8->executeString('Math.random = PHP.rand; ');
```

**Comment by virgofx** [comment]
That's awesome! Didn't know you could just assign and reference functions like that! I'm testing the snapshot via construct as well. I don't believe there will be any timing differences, but am going to check :)

**Comment by virgofx** [comment]
Snapshot via constructor now works, Timing equivalent.  I'm integrating all these changes into my V8JS PHP wrapper that handles caching. 

Latest tests passed as well (except for one skipped).  

Looks good, maybe time to merge upstream and then make a note about the `Math.random()` bug. 

One last thing: What do you think about creating a separate repository just for the stubs (Currently listed in the readme?)

Basically, this would allow developers to update composer and then have full autocomplete documentation within IDE. It could be within the same repository; however, since it's really just a development thing doesn't make sense to pull the V8JS source in a PHP application.

```
"require-dev": {
    "phpv8/v8js-stubs": "~0.4"
}
```

**Comment by stesie** [comment]
@virgofx I'm not so much an IDE user, still using good ol' Emacs :-)
... but I see that it could be helpful and that it's done for other projects as well.

Documentation actually is a good topic, since most of it + the samples are pretty outdated and not always "best practice" (in my sense) ...
